helix-core/src/date.rs

```diff
@@ -0,0 +1,217 @@
+use chrono::{Duration, NaiveDate};
+
+use std::borrow::Cow;
+
+use ropey::RopeSlice;
+
+use crate::{
+    textobject::{textobject_word, TextObject},
+    Range, Tendril,
+};
+
+// Only support formats that aren't region specific.
+static FORMATS: &[&str] = &["%Y-%m-%d", "%Y/%m/%d"];
+
+// We don't want to parse ambiguous dates like 10/11/12 or 7/8/10.
+// They must be YYYY-mm-dd or YYYY/mm/dd.
+// So 2021-01-05 works, but 2021-1-5 doesn't.
+const DATE_LENGTH: usize = 10;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct DateIncrementor {
+    pub date: NaiveDate,
+    pub range: Range,
+    pub format: &'static str,
+}
+
+impl DateIncrementor {
+    pub fn from_range(text: RopeSlice, range: Range) -> Option<DateIncrementor> {
+        // Don't increment if the cursor is one right of the date text.
+        if text.char(range.from()).is_whitespace() {
+            return None;
+        }
+
+        let range = textobject_word(text, range, TextObject::Inside, 1, true);
+        let text: Cow<str> = text.slice(range.from()..range.to()).into();
+
+        let first = text.chars().next()?;
+        let last = text.chars().next_back()?;
+
+        // Allow date strings in quotes.
+        let (range, text) = if first == last && (first == '"' || first == '\'') {
+            (
+                Range::new(range.from() + 1, range.to() - 1),
+                Cow::from(&text[1..text.len() - 1]),
+            )
+        } else {
+            (range, text)
+        };
+
+        if text.len() != DATE_LENGTH {
+            return None;
+        }
+
+        FORMATS.iter().find_map(|format| {
+            NaiveDate::parse_from_str(&text, format)
+                .ok()
+                .map(|date| DateIncrementor {
+                    date,
+                    range,
+                    format,
+                })
+        })
+    }
+
+    pub fn incremented_text(&self, amount: i64) -> Tendril {
+        let incremented_date = self.date + Duration::days(amount);
+        incremented_date.format(self.format).to_string().into()
```

Comment:
My initial thought is that the incrementing should depends on the position on the cursor, say if the cursor is in month it will increment month.

Reply:
That was my original thought as well. However, that's not so straight forward because there isn't a fixed duration for a month or a year. How would we go about incrementing the month? Just adding to the month field isn't good enough because the day of the month might not exist in the incremented month. For example how would we increment the month field of `2022-01-31` by one? Incrementing the year could also be a problem because of leap years.

Reply:
Yeah, I forgot about that.

Reply:
I ended up using https://github.com/de-vri-es/gregorian-rs

Reply:
Do we really want to use gregorian if we want to increment time later? The benefit I see here is being able to easily increment month. But later when we want to increment time we can't really do that.

Reply:
That's correct, `gregorian-rs` won't help for time. We should probably use `chrono` and implement month and year incrementing ourselves.